src/charm.py

```diff
@@ -138,38 +173,45 @@ def _publish_nrf_info_for_all_requirers(self, url: str) -> None:
             return
         self.nrf_provider.set_nrf_information_in_all_relations(url)
 
-    @property
-    def _config_file_is_written(self) -> bool:
-        """Returns whether the config file was written to the workload container.
+    def _relation_created(self, relation_name: str) -> bool:
+        """Returns whether a given Juju relation was crated.
+
+        Args:
+            relation_name (str): Relation name
 
         Returns:
-            bool: Whether the config file was written.
+            bool: Whether the relation was created.
         """
-        if not self._container.exists(f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
-            logger.info(f"Config file is not written: {CONFIG_FILE_NAME}")
-            return False
-        logger.info("Config file is written")
-        return True
+        return bool(self.model.get_relation(relation_name))
 
-    @property
-    def _database_relation_is_created(self) -> bool:
-        """Returns whether database relation is created.
+    def _push_config_file(self, content: str) -> None:
+        """Pushes config file to workload.
 
-        Returns:
-            bool: Whether database relation is created.
+        Args:
+            content: config file content
         """
-        return self._relation_created("database")
+        if not self._container.can_connect():
+            return
+        self._container.push(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}", source=content)
+        logger.info("Pushed %s config file", CONFIG_FILE_NAME)
 
-    def _relation_created(self, relation_name: str) -> bool:
-        """Returns whether a given Juju relation was crated.
+    def _database_is_available(self) -> bool:
+        """Returns True if the database is available.
 
-        Args:
-            relation_name (str): Relation name
+        Returns:
+            bool: True if the database is available.
+        """
+        return self._database.is_resource_created()
+
+    def _database_info(self) -> dict:
+        """Returns the database data.
 
         Returns:
-            bool: Whether the relation was created.
+            Dict: The database data.
         """
-        return bool(self.model.get_relation(relation_name))
+        if not self._database_is_available():
+            raise RuntimeError(f"Database `{DATABASE_NAME}` is not available")
+        return self._database.fetch_relation_data()[self._database.relations[0].id]
 
```

Comment:
It is redundant, I wanted to check that the relation data is populated, not only that the relation was created, but I realised now that the `is_resource_created` does a similar check.  I added a check to make sure the uri is in place before populating the config to avoid getting `keyerror` if the uri is late to be added to the databag for any reason. let me know if it makes sense or it just an overkill.

Reply:
Can this be resolved? @ghislainbourgeois